cryodrgn/fft.py

```diff
@@ -1,44 +1,108 @@
+from itertools import repeat
+import logging
+from multiprocessing.pool import ThreadPool as Pool
 import numpy as np
+import torch
+from torch.fft import fftshift, ifftshift, fft2, fftn, ifftn
 
-try:
-    import cupy as cp  # type: ignore
-except ImportError:
-    cp = np
 
+logger = logging.getLogger(__name__)
 
-def fft2_center(img):
-    pp = np if isinstance(img, np.ndarray) else cp
 
-    return pp.fft.fftshift(
-        pp.fft.fft2(pp.fft.fftshift(img, axes=(-1, -2))), axes=(-1, -2)
+def transform_in_chunks(
+    f, data, chunksize=None, n_workers=1, inplace=False, *args, **kwargs
+):
+    def g(src, dest, indices):
+        logger.debug(f"Transforming chunk of size {len(indices)}")
+        dest[indices, ...] = f(src[indices, ...], *args, **kwargs)
+
+    ndim = data.ndim
+    if data.ndim == 2:
+        n = 1
+        src = data[np.newaxis, ...]
+    elif data.ndim == 3:
+        n = data.shape[0]
+        src = data
+    else:
+        raise RuntimeError(f"Unsupported data ndim {ndim}")
+
+    indices = np.arange(n)
+    chunksize = min(n, chunksize) if chunksize is not None else n
+    chunk_indices = np.array_split(indices, n // chunksize)
+    n_chunks = len(chunk_indices)
+
+    if inplace:
+        dest = src
+    else:
+        if isinstance(src, np.ndarray):
+            dest = src.copy()
+        elif isinstance(src, torch.Tensor):
+            dest = src.detach().clone()
+        else:
+            raise RuntimeError("Unsupported type for src")
+
+    with Pool(n_workers) as pool:
+        pool.starmap(
+            g, zip(repeat(src, n_chunks), repeat(dest, n_chunks), chunk_indices)
+        )
+        return dest.reshape(data.shape)
+
+
+def normalize(
+    img, mean=0, std=None, std_n=None, inplace=False, chunksize=None, n_workers=1
+):
+    def _normalize(img, mean, std):
+        return (img - mean) / std
+
+    if std is None:
+        # Since std is a memory consuming process, use the first std_n samples for std determination
+        std = torch.std(img[:std_n, ...])
+
+    retVal = transform_in_chunks(
+        _normalize,
+        img,
+        chunksize=chunksize,
+        n_workers=n_workers,
+        inplace=inplace,
+        mean=mean,
+        std=std,
     )
 
+    logger.info(f"Normalized by {mean} +/- {std}")
+    return retVal
 
-def fftn_center(img):
-    pp = np if isinstance(img, np.ndarray) else cp
 
-    return pp.fft.fftshift(pp.fft.fftn(pp.fft.fftshift(img)))
+def fft2_center(img):
+    return fftshift(fft2(fftshift(img, dim=(-1, -2))), dim=(-1, -2))
```

Comment:
Great! I would do a sanity check in a unit test that the numpy and torch fft functions give you the same answer.

Reply:
I did - they're within `1e-5` of each other. I'll add the tests to the suite.

Reply:
Added in commit 7f3b87558d77091d39877540c5a6906de0ea41cc